packages/gatsby-source-filesystem/src/extend-file-node.js

```diff
@@ -0,0 +1,36 @@
+const { GraphQLString } = require(`graphql`)
+const fs = require(`fs-extra`)
+const path = require(`path`)
+
+module.exports = ({ type, getNodeAndSavePathDependency, pathPrefix = "" }) => {
+  if (type.name !== `File`) {
+    return {}
+  }
+
+  return {
+    publicURL: {
+      type: GraphQLString,
+      args: {},
+      resolve: (file, fieldArgs, context) => {
+        const details = getNodeAndSavePathDependency(file.id, context.path)
+
+        const fileName = `${file.internal.contentDigest}${details.ext}`
```

Comment:
Missing the `.` between the two strings?

Reply:
Also let's include the original file name in the new pathname like here: https://github.com/gatsbyjs/gatsby/blob/ce2f45cc27b8e0c065ca82ab82aca721c3720dd2/packages/gatsby-plugin-sharp/src/index.js#L265

Reply:
`details.ext` already contains `.`,  `details.extension` (or something like that) doesn't contain it.

Anyway this for this and rest of your definitly valid comments - this is basicly copied https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sharp/src/extend-node-type.js#L105 minus image width, height part, plus path prefix which is propably missing in code I linked, actual image processing methods (sizes, resolutions) do honor path prefix. 

I will fix those issues propably tomorrow.

Reply:
Oh haha, well please fix transformer-sharp while you're at it :-)